File: src/java/tusan/nodes/base/loops/Loop.java
package net.tutla.tums.tusan.nodes.base.loops;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.Statement;
import net.tutla.tums.tusan.nodes.expression.Expression;
import net.tutla.tums.tusan.nodes.expression.Factor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Loop extends Node {
    private Integer pos;
    private Boolean end = false;
    private Boolean run = true;

    public String as;
    public Object times;

    public Loop(Token token){
        super(token);
    }

    public Loop create(){
        if (token.type == TokenType.NUMBER){
            times = ((Double) new Expression(interpreter.currentToken).create().value).intValue();
            interpreter.expectToken(TokenType.KEYWORD, "times");
        } else if (token.type == TokenType.KEYWORD && token.value.equals("all")) {
            String target = interpreter.expectTokenClassic("KEYWORD:items|KEYWORD:characters").value;
            interpreter.expectToken(TokenType.LOGIC, "in");
            if (target.equals("characters") || target.equals("items")){
                times = new Expression(interpreter.nextToken()).create().value;
            }
        } else {
            interpreter.error("TusanError", "Tusan does not support this iterable", Arrays.asList("This is caused due to the Tusan API limitations on the parent language. To fix it use the main branch (Python) or reimplement it."));
        }
        parseAs();
        loop();
        return this;
    }

    public void loop(){
        pos = interpreter.pos;
        Boolean run = true;

        if (times instanceof Integer){
            for (int i = 0; i <= ((int) times); i++) {
                loopExecute(i);
            }
        } else {
            if (times instanceof Iterable) {
                for (Object i : (Iterable<?>) times) {
                    loopExecute(i);
                }
            } else if (times instanceof Object[]) {
                for (Object i : (Object[]) times) {
                    loopExecute(i);
                }
            } else if (times instanceof String){
                for (int i = 0; i < ((String) times).length(); i++) {
                    char c = ((String) times).charAt(i);
                    loopExecute(c);
                }
            }else {
                interpreter.error("TusanError", "Tusan does not support this iterable", null);
            }
        }
    }

    public void loopExecute(Object value){
        interpreter.data.vars.put(as, value);
        interpreter.pos = pos;
        boolean endBlock = false;

        while (!endBlock){
            Token nxt = interpreter.nextToken();
            if (nxt.type == TokenType.ENDSTRUCTURE) {
                endBlock = true;
            } else if (nxt.type == TokenType.BREAKSTRUCTURE){ // broken bcuz it won't work inside structures
                run = false;
            } else {
                if (run){
                    new Statement(nxt).create();
                }
            }
        }
    }

    public void setAs(String val){
        as = val;
        interpreter.data.vars.put(as, null);
    }

    public void parseAs(){
        if (interpreter.getNextToken().type == TokenType.KEYWORD && interpreter.getNextToken().value.equals("as")){
            interpreter.nextToken();
            Token var = interpreter.nextToken();
            if (var.type == TokenType.IDENTIFIER){
                setAs(var.value);
            }
        } else {
            setAs("loop_item");
        }
    }
}

File: src/python/src/tusan/nodes/base/loops/Loop.py
from tusk.node import Node
from tusk.token import Token
class LoopNode(Node):
    def __init__(self, token: Token):        
        self.token = token
        self.interpreter = token.interpreter
        

        self.as_ = None
        self.times = 0

    async def create(self):
        from tusk.nodes.expressions import FactorNode, ExpressionNode
        if self.token.type=="NUMBER":
            self.times = range(int((await FactorNode(self.token).create()).value))
            self.interpreter.expect_token("KEYWORD:times")
            e = await self.set_as(token=self.interpreter.get_next_token())
            await self.loop()
        elif self.token.type=="KEYWORD" and self.token.value == "all":
            loop_target_type = self.interpreter.expect_token("KEYWORD:items|KEYWORD:characters").value
            self.interpreter.expect_token("LOGIC:in")
            if loop_target_type == "characters":
                self.times = str((await ExpressionNode(self.interpreter.next_token()).create()).value)
                e = await self.set_as(token=self.interpreter.get_next_token())
                await self.loop()
            elif loop_target_type == "items":
                self.times = (await ExpressionNode(self.interpreter.next_token()).create()).value
                print(self.times,"times")
                e = await self.set_as(token=self.interpreter.get_next_token())
                await self.loop()
            else:
                raise Exception(f"loop expected token KEYWORD | NUMBER got {self.token.type}")
        return self

    async def set_as(self, token:Token=None,value=None,fallback="loop_item"):
        if value != None:
            self.interpreter.data["vars"][self.as_] = value
        else:
            if token.type == "KEYWORD" and token.value == "as":
                self.interpreter.next_token()
                var = self.interpreter.next_token()
                if var.type == "IDENTIFIER":
                    self.as_ = var.value
                    self.interpreter.data["vars"][str(var.value)] = None
            else:
                self.as_ = fallback
                self.interpreter.data["vars"][self.as_] = None
        return self

    async def loop(self):
        from tusk.nodes.statement import StatementNode
        from tusk.variable import Variable
        from tusk.nodes.base.name import setter
        pos = self.interpreter.pos
        run = True
        for i in self.times: 
            setter(self.as_,i,self.interpreter)
            end_block = False
            self.interpreter.pos = pos
            while end_block == False:
                nxt_tkn = self.interpreter.next_token()
                if nxt_tkn.type == "ENDSTRUCTURE":
                    end_block =True
                    break
                elif nxt_tkn.type == "BREAKSTRUCTURE" and nxt_tkn.value == "break":
                    run = False
                    break
                else:
                    if run == True:
                        await StatementNode(nxt_tkn).create()